src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,489 @@
+///     any custom dialogs.
+///   </para>
+/// </remarks>
+[Tool]
+public class CustomDialog : Popup, ICustomPopup
+{
+    private string windowTitle;
+
+    private bool closeHovered;
+
+    private Vector2 dragOffset;
+    private Vector2 dragOffsetFar;
+
+    private TextureButton closeButton;
+
+    private DragType dragType = DragType.None;
+
+    [Flags]
+    private enum DragType
+    {
+        None = 0,
+        Move = 1,
+        ResizeTop = 1 << 1,
+        ResizeRight = 1 << 2,
+        ResizeBottom = 1 << 3,
+        ResizeLeft = 1 << 4,
+    }
+
+    /// <summary>
+    ///   The text displayed in the window's title bar.
+    /// </summary>
+    [Export]
+    public string WindowTitle
+    {
+        get => windowTitle;
+        set
+        {
+            windowTitle = value;
+            MinimumSizeChanged();
+            Update();
+        }
+    }
+
+    /// <summary>
+    ///   If true, the user can resize the window.
+    /// </summary>
+    [Export]
+    public bool Resizable { get; set; }
+
+    /// <summary>
+    ///   If true, the window's position is clamped inside the screen so it doesn't go out of bounds.
+    /// </summary>
+    [Export]
+    public bool BoundToScreenArea { get; set; } = true;
+
+    [Export]
+    public bool ExclusiveAllowCloseOnEscape { get; private set; } = true;
+
+    public override void _EnterTree()
+    {
+        // To make popup rect readjustment react to window resizing
+        if (!GetTree().Root.IsConnected("size_changed", this, nameof(OnViewportResized)))
+            GetTree().Root.Connect("size_changed", this, nameof(OnViewportResized));
+
+        SetupCloseButton();
+        UpdateChildRects();
+    }
+
+    public override void _Notification(int what)
+    {
+        switch (what)
+        {
+            case NotificationResized:
+            {
+                UpdateChildRects();
+                break;
+            }
+
+            case NotificationVisibilityChanged:
+            {
+                if (Visible)
+                {
+                    OnShown();
+                }
+                else
+                {
+                    OnHidden();
+                }
+
+                UpdateChildRects();
+                break;
+            }
+
+            case NotificationMouseExit:
+            {
+                // Reset the mouse cursor when leaving the resizable window border.
+                if (Resizable && dragType == DragType.None)
+                {
+                    if (MouseDefaultCursorShape != CursorShape.Arrow)
+                        MouseDefaultCursorShape = CursorShape.Arrow;
+                }
+
+                break;
+            }
+        }
+    }
+
+    public override void _Draw()
+    {
+        var panel = GetStylebox("custom_panel", "WindowDialog");
+        var titleBarPanel = GetStylebox("custom_titlebar", "WindowDialog");
+        var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
```

Comment:
Should we maybe cache these constant retrieves? Each one of them causes a call into the native Godot code, which is probably many times slower than calling C# code.